bdb-ip.md

```diff
@@ -0,0 +1,437 @@
+
+- `unspent=true`: Return all outputs that have not been spent or "unspent"
+- `unspent=false`: Return all outputs that have not been unspent or "spent"
+(bit of a stretch to get here in your mind, try it)
+
+
+##### Proposal 1
+
+- Rename `unspent` to `spent`
+* If `spent=true`, response is an array of all spent outputs associated with
+  that public key
+* If `spent=false`, response is an array of all *NOT YET* spent (or "unspent"
+  outputs associated with that public key
+* If no unspent= filter, response is an array of all outputs associated with
+  that public key (spent and unspent)
+
+
+Pro:
+
+- Resolves mind-stretch exercise
+- Resolves unexpected result when querying the API
+
+Contra:
+
+- Lots of changes to be applied to core, drivers, etc.
+- `unspent` is a word of daily communication that we're all used to
+`spent=false` not
+
+
+##### Proposal 2
+
+Based on what @ttmc suggested [here](https://github.com/bigchaindb/bigchaindb/issues/1214#issuecomment-282240980).
+
+* If unspent=true, response is an array of all unspent outputs associated with
+  that public key
+* If unspent=false, response is an array of all spent outputs associated with
+  that public key
+* If no unspent= filter, response is an array of all outputs associated with
+  that public key (spent and unspent)
+
+
+Pro:
+
+- Resolves unexpected result when querying the API
+- Less changes to be applied to core, drivers, etc. as resolvement 1
+
+Contra:
+
+- Doesn't resolves mind-stretch exercise
+
+
+##### Proposal 3
+
+Based on what @vrde and @r-marques suggested
+[here](https://github.com/bigchaindb/bigchaindb-driver/issues/275) and
+[here](https://github.com/bigchaindb/bigchaindb/issues/1214#issuecomment-290400956):
+
+- Rename `unspent` to either `unspent_only` or `exclude_spent` or
+`exclude_unspent`.
+
+
+Pro:
+
+- Expected result = actual result
+
+Contra:
+
+- Doesn't resolve issue properly as the endpoint should:
+    - Return an array with both spent and unspent endpoint
+    - Return an array with only spent outputs
+    - Return an array with only unspent outputs
+- Doesn't resolves mind-stretch exercise
+
+
+##### Personal favorite: Proposal 1
+
+
+#### [/statuses?tx_id needs to return status invalid](https://github.com/bigchaindb/bigchaindb/issues/1039)
+
+Contra:
+
+- [Is VERY difficult to implement in a consistent manner as we don't store invalid
+transactions and since only blocks are marked invalid, transactions are not](https://github.com/bigchaindb/bigchaindb/issues/1039#issuecomment-290424788)
+- There is better ways to serve this use case:
+    - Tail BDB's logs (generally: provide better debugging capabilities)
+    - Listen to the BDB event stream API
+    - Event stream API for rejected transactions?
+
+##### Proposal 1
+
+- Generalize use case (e.g. As a BigchainDB user, I want to know when my
+  transactions get rejected by a BigchainDB node) and open new ticket
+- Close #1039
+
+
+##### Personal favorite: Proposal 1 lol
```

Comment:
This is a hard problem indeed and #1039 has a good discussion about the issue.
I agree that if it is a feature that we want we can create a new issue for it (and add it to the non-breaking changes milestone?)

In the meantime the api validates the transaction when it is pushed to a node and if invalid it sends a message back to client with the reason.

If the transaction is accepted by a node and ends up being invalid (double spend attempt?) the status endpoint will return a `404`. The client can always resend the transaction and get back a message with why the transaction is invalid.